File: 143_reorderList.cpp
// Time Complexity : O(N)
// Space Complexity :O(1)

class Solution {
public:
    void reorderList(ListNode* head) {
        if(head == NULL)
            return;
        ListNode* fast = head;
        ListNode* slow = head;
        while(fast != NULL && fast->next != NULL){
            fast = fast->next->next;
            slow = slow->next;
        }
        fast = reverseList(slow->next);
        slow->next = NULL;
        slow = head;
        while(fast != NULL){
            ListNode* temp = slow->next;
            slow->next = fast;
            fast = fast->next;
            slow->next->next = temp;
            slow = temp;
        }
    }
    
    ListNode* reverseList(ListNode* head){
        if(head == NULL)
            return head;
        ListNode* prev = NULL;
        ListNode* curr = head;
        ListNode* fast = head->next;
        while(fast != NULL){
            curr->next = prev;
            prev = curr;
            curr = fast;
            fast = fast->next;
        }
        curr->next = prev;
        return curr;
    }
};

File: _143_ReorderList.java
// space complexity - o(1)
      //time - o(n) where n is length of linked list
      //Ran on Leetcode successfully : Yes
      // Problem faced  : No
// Iterative approach : getLengthOfList, breakAndGetSecondHalfHead, reverseSecondHalf , mergeTwoList
class Solution {

    public void reorderList(ListNode head) {
        if(head == null || head.next == null) return;
        int len = getLengthOfList(head);
        ListNode secondHead = breakAndGetSecondHalfHead(head, len);
        secondHead = reverseSecondHalf(secondHead);
        mergeTwoList(head, secondHead);
    }

    public void mergeTwoList(ListNode headA, ListNode headB){

        ListNode dummy = new ListNode(Integer.MIN_VALUE);

        while(headA != null || headB != null){

            if(headA != null){
                dummy.next = headA;
                headA = headA.next;
                dummy = dummy.next;
                dummy.next = null;
            }
            if(headB != null){
                dummy.next = headB;
                headB = headB.next;
                dummy = dummy.next;
                dummy.next = null;
            }
        }
    }

   public int getLengthOfList(ListNode head){
        int res = 0 ;
        while(head != null){
            head = head.next;
            res++;
        }
        return res;
    }

    public ListNode breakAndGetSecondHalfHead(ListNode head, int len){
        int res = len/2 ;
        while(res > 1){
            head = head.next;
            res--;
        }
        ListNode headB = head.next;
        head.next = null;
        return headB;
    }

    public ListNode reverseSecondHalf(ListNode head){
        ListNode prev = null;
        ListNode curr = head;

        while(curr != null){
            ListNode next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }
